pkg/errctx/context.go

```diff
@@ -0,0 +1,152 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package errctx
+
+import (
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/pkg/errno"
+	"github.com/pingcap/tidb/pkg/util/intest"
+)
+
+// Level defines the behavior for each error
+type Level uint8
+
+const (
+	// LevelError means the error will be returned
+	LevelError Level = iota
+	// LevelWarn means it will be regarded as a warning
+	LevelWarn
+	// LevelIgnore means the error will be ignored
+	LevelIgnore
+)
+
+// Context defines how to handle an error
+type Context struct {
+	levelMap        [2]Level
+	appendWarningFn func(err error)
+}
+
+// WithStrictErrGroupLevel makes the context to return the error directly for any kinds of errors.
+func (ctx *Context) WithStrictErrGroupLevel() Context {
+	newCtx := Context{
+		appendWarningFn: ctx.appendWarningFn,
+	}
+
+	return newCtx
+}
+
+// WithErrGroupLevel sets a `Level` for an `ErrGroup`
+func (ctx *Context) WithErrGroupLevel(eg ErrGroup, l Level) Context {
+	newCtx := Context{
+		levelMap:        ctx.levelMap,
+		appendWarningFn: ctx.appendWarningFn,
+	}
+	newCtx.levelMap[eg] = l
+
+	return newCtx
+}
+
+// AppendWarning appends the error to warning. If the inner `appendWarningFn` is nil, do nothing.
+func (ctx *Context) AppendWarning(err error) {
+	intest.Assert(ctx.appendWarningFn != nil)
+	if fn := ctx.appendWarningFn; fn != nil {
+		// appendWarningFn should always not be nil, check fn != nil here to just make code safe.
+		fn(err)
+	}
+}
+
+// HandleError handles the error according to the context.
+//  1. If the `internalErr` is not `"pingcap/errors".Error`, or the error code is not defined in the `errGroupMap`, or the error
+//     level is set to `LevelError`(0), the `err` will be returned directly.
+//  2. If the error level is set to `LevelWarn`, the `warnErr` will be appended as a warning.
+//  3. If the error level is set to `LevelIgnore`, this function will return a `nil`.
+//
+// In most cases, these three should be the same. If there are many different kinds of error internally, but they are expected
+// to give the same error to users, the `err` can be different form `internalErr`. Also, if the warning is expected to be
+// different from the initial error, you can also use the `warnErr` argument.
+//
+// TODO: would it be better to use `HandleError(err error, ...error)`? As the second and third arguments are the same with
+// the first one is most cases.
+// TODO: is it good to give an error code for internal only errors? Or should we use another way to distinguish different
+// group of errors?
+// TODO: both `types.Context` and `errctx.Context` can handle truncate error now. Refractor them.
+func (ctx *Context) HandleError(internalErr error, err error, warnErr error) error {
```

Comment:
This method has three arguments but the second and the third is the same with the first one in most cases. Can we rename this method to `HandleErrorWithAlias` and then make `HandleError` with only one parameter and same with `HandleErrorWithAlias(err, err, err)`